### PR TITLE
upgrade libp2p to v0.53.*

### PR DESCRIPTION
**Pull request by divagant-martian**
## Issue Addressed

Upgrades libp2p to v0.53. See https://github.com/libp2p/rust-libp2p/releases/tag/libp2p-v0.53.0 and the subsequent `.1` release

## Proposed Changes

This adds metrics for the transports directly from libp2p with a better granularity. It also changes the way handlers are closed. We no longer have a way to force closing a handler, instead libp2p relies on checking established substreams. Rest of the changes are not too relevant

## Additional Info

na

### CODE REVIEW

**Comment by divagant-martian** [comment]
@mxinden I'm curious about the bandwidth metrics here. I'm getting for example
```
libp2p_bandwidth_bytes_total{protocols="/ip4/tcp",direction="Inbound"} 2967
libp2p_bandwidth_bytes_total{protocols="/ip4/udp/quic-v1/p2p",direction="Outbound"} 0
libp2p_bandwidth_bytes_total{protocols="/ip4/tcp/p2p",direction="Outbound"} 33160
libp2p_bandwidth_bytes_total{protocols="/ip4/tcp",direction="Outbound"} 2845
libp2p_bandwidth_bytes_total{protocols="/ip4/tcp/p2p",direction="Inbound"} 1661330
libp2p_bandwidth_bytes_total{protocols="/ip4/udp/quic-v1/p2p",direction="Inbound"} 0
```

What's the difference between 
- `libp2p_bandwidth_bytes_total{protocols="/ip4/tcp/p2p",direction="Outbound"}`
- `libp2p_bandwidth_bytes_total{protocols="/ip4/tcp",direction="Outbound"}` ?

**Comment by mxinden** [comment]
> What's the difference between
> 
> 
>   * `libp2p_bandwidth_bytes_total{protocols="/ip4/tcp/p2p",direction="Outbound"}`

Here either a `NetworkBehaviour` or `Swarm::dial` dialed a ipv4 tcp with a `/p2p/SOME_PEER_ID`.

>   * `libp2p_bandwidth_bytes_total{protocols="/ip4/tcp",direction="Outbound"}` ?

Here either a `NetworkBehaviour` or `Swarm::dial` dialed a ipv4 tcp without a `/p2p/SOME_PEER_ID`.

Not quite ideal. We could add the peer id to the multiaddr in case it doesn't have one once the dial succeeded, and thus always have a `/ip4/tcp/p2p`.

https://github.com/libp2p/rust-libp2p/blob/3b6b74db7fdb252103699efcbc85cb11605ea893/misc/metrics/src/bandwidth.rs#L92-L94

File: beacon_node/lighthouse_network/Cargo.toml [hunk 1]
```diff
@@ -39,17 +39,17 @@ directory = { workspace = true }
...
 strum = { workspace = true }
 superstruct = { workspace = true }
-prometheus-client = "0.21.0"
+prometheus-client = "0.22.0"
 unused_port = { workspace = true }
 delay_map = { workspace = true }
 void = "1"
-libp2p-quic= { version = "0.9.2", features=["tokio"]}
-libp2p-mplex = "0.40.0"
+libp2p-quic= { version = "0.10.1", features=["tokio"]}
```
Comment (jxs) [inline beacon_node/lighthouse_network/Cargo.toml]: since are on this, I think we can import quic via libp2p, adding the `quic` feature and importing `libp2p::quic` can you do it Diva?

File: beacon_node/lighthouse_network/src/service/utils.rs [hunk 2]
```diff
@@ -53,44 +52,35 @@ pub fn build_transport(
     // yamux config
     let mut yamux_config = yamux::Config::default();
     yamux_config.set_window_update_mode(yamux::WindowUpdateMode::on_read());
-
     // Creates the TCP transport layer
-    let (tcp, tcp_bandwidth) =
```
Comment (jxs) [inline beacon_node/lighthouse_network/src/service/utils.rs:58]: I think we can now incorporate Max's changes presented on https://github.com/sigp/lighthouse/pull/4695. If you feel like it's too much work don't worry, I can do it on a subsequent PR

**Review by jxs — APPROVED** [review]
Thanks Diva!

File: beacon_node/lighthouse_network/src/service/utils.rs (hunk 2 above)
Comment (divagant-martian) [reply beacon_node/lighthouse_network/src/service/utils.rs:58]: ..this is already done in this pr @jxs

**Comment by divagant-martian** [comment]
thanks @jxs. However this is not yet properly tested (thus the draft status). This upgrade also affects the way libp2p handles closing handlers so it would be good if @AgeManning gives a look at the RPC handler changes. I went with a simple approach that I think should work but I'd like his opinion on this

File: beacon_node/lighthouse_network/src/service/utils.rs (hunk 2 above)
Comment (divagant-martian) [reply beacon_node/lighthouse_network/src/service/utils.rs:58]: I looked into this but it looks less clean imo to jam the tcp configuration in a closure when it's already well contained in a function (`build_transport`). In that PR Thomas notes as well this. Tbh I don't see the benefit

File: beacon_node/lighthouse_network/Cargo.toml (hunk 1 above)
Comment (divagant-martian) [reply beacon_node/lighthouse_network/Cargo.toml]: done in 6a8c835df20ffbbea1613fda7ab53d758963dd17

File: beacon_node/lighthouse_network/src/config.rs [hunk 3]
```diff
@@ -518,7 +518,7 @@ pub fn gossipsub_config(
         .validation_mode(gossipsub::ValidationMode::Anonymous)
         .duplicate_cache_time(DUPLICATE_CACHE_TIME)
         .message_id_fn(gossip_message_id)
-        .fast_message_id_fn(fast_gossip_message_id)
+        .message_id_fn(fast_gossip_message_id)
```
Comment (AgeManning) [inline beacon_node/lighthouse_network/src/config.rs]: This change doesn't look right. 

I think the fast message id function in gossipsub was removed as we decided it probably wasn't giving us too many gains on the compression. 

I think we can remove this line and the entire fast message id function (in Lighthouse). In this current form, I think we'll mismatch with all client teams on the message-id function, which won't be fun.

**Review by AgeManning — COMMENTED** [review]
With the RPC things I'm not sure this will work. 

Can we no longer close connections at the handler level? 

We designed lighthouse such that it was ultimately the RPC connection handler that closes connections to peers, so that we could send a goodbye message to them before we drop the connection. 

Unless libp2p have changed some things, I recall that the `connection_keep_alive` is grouped with all other behaviours. So even if the RPC says the keep_alive is false, the connection won't close if gossipsub's keep_alive is true or identify or ping etc. 

I think we might need to explicitly inform the behaviour to drop the connection if we can't do it in the handler any more. 

There is a chance I'm wrong about this, will have to double check.

**Comment by jxs** [comment]
> Can we no longer close connections at the handler level?

yeah, see https://github.com/libp2p/rust-libp2p/pull/4755

**Comment by divagant-martian** [comment]
yeah I mentioned this in the description
> We no longer have a way to force closing a handler, instead libp2p relies on checking established substreams.

Which is a bit annoying and somewhat reverts to the state of things a couple of years back. There is a way to ignore for keep alive, where our problematic contender would be gossipsub. I'll look into that. Otherwise we would need to bubble up a new error kind since right now reporting is done on a substream level and the couple of errors modified do not fit in this model. I'll try to check the first approach before going with modifying the handler errors

File: beacon_node/lighthouse_network/src/config.rs (hunk 3 above)
Comment (divagant-martian) [reply beacon_node/lighthouse_network/src/config.rs]: oh you are completely right, this was definitely absentmindedly done

Comment (divagant-martian) [reply beacon_node/lighthouse_network/src/config.rs]: done in https://github.com/sigp/lighthouse/pull/4935/commits/a63244ac69b44d668402474aa2a6cdd6918c7432

**Comment by divagant-martian** [comment]
Well, that idea died rather quickly. 

`ignore_for_keep_alive` is available at the stream level so there is no way to do it per sub-behaviour. 

@mxinden would it make sense to allow the `NetworkBehaviour` derive to tag sub-behaviours with custom `keep-alive` overrides?. In LH's case the override would be constant function `false` for each one except the rpc handler. Mearly a suggestion, since the new state is less flexible than previous one.

For this PR tho, I'll go with sending up the errors to the behaviour for it to close the connection. This approach is more resilient to changes in libp2p internals anyway

File: beacon_node/lighthouse_network/src/service/mod.rs [hunk 4]
```diff
@@ -1431,6 +1451,11 @@ impl<AppReqId: ReqId, TSpec: EthSpec> Network<AppReqId, TSpec> {
                 };
                 self.build_response(id, peer_id, response)
             }
+            HandlerEvent::Close(_) => {
+                let _ = self.swarm.disconnect_peer_id(peer_id);
```
Comment (divagant-martian) [inline beacon_node/lighthouse_network/src/service/mod.rs:1455]: Sorry for the noise @mxinden 

I think the new API does not fully allow the previous behavior. 

When emitting an event from the handler, this is received on `swarm.poll_next_unpin` without the information of the `connection_id`, so that this line could be `close_connection` instead. 

For LH this is good enough since we allow only one connection per peer, but it seems like a slight regression in connection management in libp2p. 

Now, if there is a way to achieve this "the correct way" (closing the specific connection) and this is an oversight on my part, happy to hear and use the alternative in this pr

Comment (mxinden) [reply beacon_node/lighthouse_network/src/service/mod.rs:1455]: I am missing context here. What is `HandlerEvent::Close`? I can only find this type alias:

``` rust
/// Events the handler emits to the behaviour.
pub type HandlerEvent<Id, T> = Result<RPCReceived<Id, T>, HandlerErr<Id>>;
```

I am assuming that this is `std`'s `Result` which does not have a `Close` variant.

**Comment by mxinden** [comment]
> @mxinden would it make sense to allow the `NetworkBehaviour` derive to tag sub-behaviours with custom `keep-alive` overrides?. In LH's case the override would be constant function `false` for each one except the rpc handler. Mearly a suggestion, since the new state is less flexible than previous one.

Can you summarize the motivation behind this change? Note that `connection_keep_alive` is a method on `ConnectionHandler` and not on `NetworkBehaviour`. Also note, that you can set `SwarmConfig::with_idle_connection_timeout` to a very high value to effectively disable any closes due to `KeepAlive::False`.

File: beacon_node/lighthouse_network/src/rpc/handler.rs [hunk 5]
```diff
@@ -51,7 +50,12 @@ impl SubstreamId {
 type InboundSubstream<TSpec> = InboundFramed<Stream, TSpec>;
 
 /// Events the handler emits to the behaviour.
-pub type HandlerEvent<Id, T> = Result<RPCReceived<Id, T>, HandlerErr<Id>>;
+#[derive(Debug)]
+pub enum HandlerEvent<Id, T: EthSpec> {
+    Ok(RPCReceived<Id, T>),
+    Err(HandlerErr<Id>),
+    Close(RPCError),
+}
```
Comment (divagant-martian) [inline beacon_node/lighthouse_network/src/rpc/handler.rs:58]: @mxinden  are you missing a pull?

File: beacon_node/lighthouse_network/src/service/mod.rs (hunk 4 above)
Comment (divagant-martian) [reply beacon_node/lighthouse_network/src/service/mod.rs:1455]: I think what you are missing is a pull max. This type is no longer a result

**Comment by divagant-martian** [comment]
> > @mxinden would it make sense to allow the `NetworkBehaviour` derive to tag sub-behaviours with custom `keep-alive` overrides?. In LH's case the override would be constant function `false` for each one except the rpc handler. Mearly a suggestion, since the new state is less flexible than previous one.
> 
> Can you summarize the motivation behind this change? Note that `connection_keep_alive` is a method on `ConnectionHandler` and not on `NetworkBehaviour`. Also note, that you can set `SwarmConfig::with_idle_connection_timeout` to a very high value to effectively disable any closes due to `KeepAlive::False`.

Deriving `NetworkBehaviour` for a struct that combines sub-behaviours also builds the associated handler. The issue here is that we don't have access to the `connection_keep_alive` function of the _generated_ handler. The simplest entry path for this that I see is the `NetworkBehaviour` derive

Also, the issue is not to prevent closing, but to trigger it

File: beacon_node/lighthouse_network/src/rpc/handler.rs [hunk 6]
```diff
@@ -831,7 +844,7 @@ where
                 && self.events_out.is_empty()
                 && self.dial_negotiated == 0
             {
-                return Poll::Ready(ConnectionHandlerEvent::Close(RPCError::Disconnected));
+                // return Poll::Ready(ConnectionHandlerEvent::Close(RPCError::Disconnected));
```
Comment (AgeManning) [inline beacon_node/lighthouse_network/src/rpc/handler.rs]: We want to return the close result here too right?

Why the comment?

**Review by AgeManning — COMMENTED** [review]
Yeah this is a nice soln! 

Just a single comment
